This notebook (and the slides from lecture 8) will help you go straight from training a model in Colab to deploying it in a webpage with TensorFlow.js - without having to leave the browser.

Configure this notebook to work with your GitHub account by populating these fields.

In [0]:
!pip install tensorflowjs

    100% |████████████████████████████████| 51kB 3.3MB/s 
keras 2.2.2 has requirement keras-applications==1.0.4, but you'll have keras-applications 1.0.6 which is incompatible.
keras 2.2.2 has requirement keras-preprocessing==1.0.2, but you'll have keras-preprocessing 1.0.5 which is incompatible.
  Found existing installation: Keras-Applications 1.0.4
    Uninstalling Keras-Applications-1.0.4:
      Successfully uninstalled Keras-Applications-1.0.4
  Found existing installation: Keras-Preprocessing 1.0.2
    Uninstalling Keras-Preprocessing-1.0.2:
      Successfully uninstalled Keras-Preprocessing-1.0.2


In [0]:
# your github username
USER_NAME = "Micky-angela" 

# the email associated with your commits
# (may not matter if you leave it as this)
USER_EMAIL = "meiqichi@hotmail.com" 

# the user token you've created (see the lecture 8 slides for instructions)
TOKEN = "e038a21219e8ac4a51d65af8e81f1719bf5624dc" 

# site name
# for example, if my user_name is "foo", then this notebook will create
# a site at https://foo.github.io/hw4/
SITE_NAME = "hw4"

Next, run this cell to configure git.

In [0]:
!git config --global user.email {USER_NAME}
!git config --global user.name  {USER_EMAIL}

Clone your GitHub pages repo (see the lecture 8 slides for instructions on how to create one).

In [0]:
import os
repo_path = USER_NAME + '.github.io'
if not os.path.exists(os.path.join(os.getcwd(), repo_path)):
  !git clone https://{USER_NAME}:{TOKEN}@github.com/{USER_NAME}/{USER_NAME}.github.io

In [0]:
os.chdir(repo_path)
!git pull

From https://github.com/Micky-angela/Micky-angela.github.io
   1109790..4d531da  master     -> origin/master
Already up to date.


Create a folder for your site.

In [0]:
project_path = os.path.join(os.getcwd(), SITE_NAME)
if not os.path.exists(project_path): 
  os.mkdir(project_path)
os.chdir(project_path)

These paths will be used by the converter script.

In [0]:
# DO NOT MODIFY
MODEL_DIR = os.path.join(project_path, "model_js")
if not os.path.exists(MODEL_DIR):
  os.mkdir(MODEL_DIR)

As an example, we will create and vectorize a few documents. (Check out https://www.gutenberg.org/ for a bunch of free e-books.)

In [0]:
import urllib.request
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [0]:
# define a function to find the position of n+1 occurence of needle in haystack
def findnth(haystack, needle, n):
    parts= haystack.split(needle, n+1)
    if len(parts)<=n+1:
        return -1
    return len(haystack)-len(parts[-1])-len(needle)

In [0]:
# define function to read text on url, eliminate unwanted characters and return list of sentences
def read_data(url):
    fp = urllib.request.urlopen(url)
    mybytes = fp.read()
    mystr = mybytes.decode("utf8")
    fp.close()
    mystr = mystr.replace('\r\n', '')
    num = findnth(mystr, '*', 5)
    mystr = mystr[num+1:]
    tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')
    book = tokenizer.tokenize(mystr)
    return book

In [0]:
# get train data
book1 = read_data('https://www.gutenberg.org/files/1342/1342.txt')[1:1001] # Pride and Prejudice
book2 = read_data('https://www.gutenberg.org/files/1661/1661.txt')[1:1001] # The Adventure of Sherlock Holmes
book3 = read_data('https://www.gutenberg.org/files/103/103.txt')[1:1001]   # Around the World in 80 Days

In [0]:
# get test data
book1_test = read_data('https://www.gutenberg.org/files/1342/1342.txt')[1001:1201] # Pride and Prejudice
book2_test = read_data('https://www.gutenberg.org/files/1661/1661.txt')[1001:1201] # The Adventure of Sherlock Holmes
book3_test = read_data('https://www.gutenberg.org/files/103/103.txt')[1001:1201]   # Around the World in 80 Days

In [0]:
x_train = book1 + book2 + book3
y_train = [0]*1000 + [1]*1000 + [2]*1000 # Indicating which book each sentence is from

In [0]:
x_test = book1_test + book2_test + book3_test
y_test = [0]*200 + [1]*200 + [2]*200 # Indicating which book each sentence is from

Tokenize the documents, create a word index (word -> number).

In [0]:
max_len = 40
num_words = 100000
from keras.preprocessing.text import Tokenizer
# Fit the tokenizer on the training data
t = Tokenizer(num_words=num_words)
t.fit_on_texts(x_train)

Using TensorFlow backend.


In [0]:
print(t.word_index)

{'the': 1, 'and': 2, 'to': 3, 'of': 4, 'a': 5, 'i': 6, 'in': 7, 'he': 8, 'was': 9, 'it': 10, 'his': 11, 'that': 12, 'you': 13, 'with': 14, 'is': 15, 'not': 16, 'at': 17, 'as': 18, 'her': 19, 'had': 20, 'for': 21, 'but': 22, 'be': 23, 'mr': 24, 'she': 25, 'which': 26, 'on': 27, 'have': 28, 'by': 29, 'my': 30, 'said': 31, 'him': 32, 'from': 33, 'this': 34, 'were': 35, 'would': 36, 'fogg': 37, 'so': 38, 'an': 39, 'they': 40, 'your': 41, 'all': 42, 'are': 43, 'who': 44, 'no': 45, 'me': 46, 'what': 47, 'very': 48, 'when': 49, 'do': 50, 'if': 51, 'will': 52, 'one': 53, 'been': 54, 'there': 55, 'then': 56, 'we': 57, 'or': 58, 'could': 59, 'passepartout': 60, 'them': 61, 'am': 62, 'their': 63, 'bingley': 64, 'more': 65, 'well': 66, 'much': 67, 'out': 68, 'miss': 69, 'up': 70, 'upon': 71, 'only': 72, 'did': 73, 'man': 74, 'sir': 75, 'some': 76, 'such': 77, 'than': 78, 'has': 79, 'good': 80, 'phileas': 81, 'must': 82, 'into': 83, 'bennet': 84, 'about': 85, 'now': 86, 'know': 87, 'how': 88, 'darc

Here's how we vectorize a document.

In [0]:
vectorized = t.texts_to_sequences([book1[10]])
print(vectorized)

[[15, 8, 511, 58, 675, 189]]


Apply padding if necessary.

In [0]:
from keras.preprocessing.sequence import pad_sequences
padded = pad_sequences(vectorized, maxlen=max_len, padding='post')

In [0]:
print(padded)

[[ 15   8 511  58 675 189   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0]]


We will save the word index in metadata. Later, we'll use it to convert words typed in the browser to numbers for prediction.

In [0]:
metadata = {
  'word_index': t.word_index,
  'max_len': max_len,
  'vocabulary_size': num_words,
}

Define a model.

In [0]:
embedding_size = 8
n_classes = 3
epochs = 10

import keras
model = keras.Sequential()
model.add(keras.layers.Embedding(num_words, embedding_size, input_shape=(max_len,)))
model.add(keras.layers.LSTM(16, return_sequences=True))
model.add(keras.layers.LSTM(16))
model.add(keras.layers.Dense(3, activation='softmax'))
model.compile('adam', 'sparse_categorical_crossentropy', metrics=['accuracy'])
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 40, 8)             800000    
_________________________________________________________________
lstm_1 (LSTM)                (None, 40, 16)            1600      
_________________________________________________________________
lstm_2 (LSTM)                (None, 16)                2112      
_________________________________________________________________
dense_1 (Dense)              (None, 3)                 51        
Total params: 803,763
Trainable params: 803,763
Non-trainable params: 0
_________________________________________________________________


In [0]:
early_stopping = keras.callbacks.EarlyStopping(monitor='val_loss',
                                               min_delta=0.001,
                                               patience=0,
                                               verbose=1)

Prepare some training data.

In [0]:
x_train = t.texts_to_sequences(x_train)
x_train = pad_sequences(x_train, maxlen=max_len, padding='post')
print(x_train)

[[  30  142   24 ...    0    0    0]
 [  84  137   12 ...    0    0    0]
 [  22   10   15 ...    0    0    0]
 ...
 [ 952   12    1 ...    0    0    0]
 [1411    3    1 ...    0    0    0]
 [  37    2   11 ...    0    0    0]]


In [0]:
x_test = t.texts_to_sequences(x_test)
x_test = pad_sequences(x_test, maxlen=max_len, padding='post')

In [0]:
model.fit(x_train, y_train, epochs=epochs, validation_data=(x_test, y_test),
                    callbacks=[early_stopping])

Train on 3000 samples, validate on 600 samples
Epoch 1/10
3000/3000 [==============================] - 7s 2ms/step - loss: 1.0959 - acc: 0.3617 - val_loss: 1.0863 - val_acc: 0.3783
Epoch 2/10
3000/3000 [==============================] - 6s 2ms/step - loss: 0.9769 - acc: 0.4743 - val_loss: 0.8764 - val_acc: 0.5300
Epoch 3/10
3000/3000 [==============================] - 6s 2ms/step - loss: 0.7020 - acc: 0.6670 - val_loss: 0.7607 - val_acc: 0.6500
Epoch 4/10
3000/3000 [==============================] - 6s 2ms/step - loss: 0.4210 - acc: 0.8467 - val_loss: 0.7022 - val_acc: 0.7333
Epoch 5/10
3000/3000 [==============================] - 6s 2ms/step - loss: 0.2571 - acc: 0.9067 - val_loss: 0.7311 - val_acc: 0.7450
Epoch 00005: early stopping


Demo using the model to make predictions.

In [0]:
test_example = "Left Munich at 8:35 P. M., on 1st May, arriving at Vienna early next morning."
x_t = t.texts_to_sequences([test_example])
x_t = pad_sequences(x_t, maxlen=max_len, padding='post')
print(x_t)

[[ 132   17 3063 7559  655  383   27  109   17  706  266  160    0    0
     0    0    0    0    0    0    0    0    0    0    0    0    0    0
     0    0    0    0    0    0    0    0    0    0    0    0]]


In [0]:
preds = model.predict(x_t)
print(preds)
import numpy as np
print(np.argmax(preds))

[[0.27596748 0.09467001 0.6293625 ]]
2


In [0]:
model.evaluate(x_test, y_test)

600/600 [==============================] - 0s 311us/step


[0.7311172103881836, 0.7450000007947286]

So the accuracy of this model on test set is 74.5%.

Convert the model

In [0]:
import json
import tensorflowjs as tfjs

metadata_json_path = os.path.join(MODEL_DIR, 'metadata.json')
json.dump(metadata, open(metadata_json_path, 'wt'))
tfjs.converters.save_keras_model(model,                                                                MODEL_DIR)
print('\nSaved model artifcats in directory: %s' % MODEL_DIR)


Saved model artifcats in directory: /content/Micky-angela.github.io/hw4/model_js


Write an index.html and an index.js file configured to load our model.

In [0]:
index_html = """
<!doctype html>

<body>
  <style>
    #textfield {
      font-size: 120%;
      width: 60%;
      height: 200px;
    }
  </style>
  <h1>
    Title
  </h1>
  <hr>
  <div class="create-model">
    <button id="load-model" style="display:none">Load model</button>
  </div>
  <div>
    <div>
      <span>Vocabulary size: </span>
      <span id="vocabularySize"></span>
    </div>
    <div>
      <span>Max length: </span>
      <span id="maxLen"></span>
    </div>
  </div>
  <hr>
  <div>
    <select id="example-select" class="form-control">
      <option value="example1">Pride and Prejudice</option>
      <option value="example2">The Adventure of Sherlock Holmes</option>
      <option value="example3">Around the World in 80 Days</option>
    </select>
  </div>
  <div>
    <textarea id="text-entry"></textarea>
  </div>
  <hr>
  <div>
    <span id="status">Standing by.</span>
  </div>

  <script src='https://cdn.jsdelivr.net/npm/@tensorflow/tfjs/dist/tf.min.js'></script>
  <script src='index.js'></script>
</body>
"""

In [0]:
index_js = """
const HOSTED_URLS = {
  model:
      'model_js/model.json',
  metadata:
      'model_js/metadata.json'
};

const examples = {
  'example1':
      'I cannot forget the follies and vices of othersso soon as I ought, nor their offenses against myself.',
  'example2':
      'I sat downbeside the fire, however, with the intention of awaiting him,however long he might be.',
  'example3':
      'You know that a visa is useless, and that no passportis required?'      
};

function status(statusText) {
  console.log(statusText);
  document.getElementById('status').textContent = statusText;
}

function showMetadata(metadataJSON) {
  document.getElementById('vocabularySize').textContent =
      metadataJSON['vocabulary_size'];
  document.getElementById('maxLen').textContent =
      metadataJSON['max_len'];
}

function settextField(text, predict) {
  const textField = document.getElementById('text-entry');
  textField.value = text;
  doPredict(predict);
}

function setPredictFunction(predict) {
  const textField = document.getElementById('text-entry');
  textField.addEventListener('input', () => doPredict(predict));
}

function disableLoadModelButtons() {
  document.getElementById('load-model').style.display = 'none';
}

function doPredict(predict) {
  const textField = document.getElementById('text-entry');
  const result = predict(textField.value);
  score_string = "Class scores: ";
  for (var x in result.score) {
    score_string += x + " ->  " + result.score[x].toFixed(3) + ", "
  }
  //console.log(score_string);
  status(
      score_string + ' elapsed: ' + result.elapsed.toFixed(3) + ' ms)');
}

function prepUI(predict) {
  setPredictFunction(predict);
  const testExampleSelect = document.getElementById('example-select');
  testExampleSelect.addEventListener('change', () => {
    settextField(examples[testExampleSelect.value], predict);
  });
  settextField(examples['example1'], predict);
}

async function urlExists(url) {
  status('Testing url ' + url);
  try {
    const response = await fetch(url, {method: 'HEAD'});
    return response.ok;
  } catch (err) {
    return false;
  }
}

async function loadHostedPretrainedModel(url) {
  status('Loading pretrained model from ' + url);
  try {
    const model = await tf.loadModel(url);
    status('Done loading pretrained model.');
    disableLoadModelButtons();
    return model;
  } catch (err) {
    console.error(err);
    status('Loading pretrained model failed.');
  }
}

async function loadHostedMetadata(url) {
  status('Loading metadata from ' + url);
  try {
    const metadataJson = await fetch(url);
    const metadata = await metadataJson.json();
    status('Done loading metadata.');
    return metadata;
  } catch (err) {
    console.error(err);
    status('Loading metadata failed.');
  }
}

class Classifier {

  async init(urls) {
    this.urls = urls;
    this.model = await loadHostedPretrainedModel(urls.model);
    await this.loadMetadata();
    return this;
  }

  async loadMetadata() {
    const metadata =
        await loadHostedMetadata(this.urls.metadata);
    showMetadata(metadata);
    this.maxLen = metadata['max_len'];
    console.log('maxLen = ' + this.maxLen);
    this.wordIndex = metadata['word_index']
  }

  predict(text) {
    // Convert to lower case and remove all punctuations.
    const inputText =
        text.trim().toLowerCase().replace(/(\.|\,|\!)/g, '').split(' ');
    // Look up word indices.
    const inputBuffer = tf.buffer([1, this.maxLen], 'float32');
    for (let i = 0; i < inputText.length; ++i) {
      const word = inputText[i];
      inputBuffer.set(this.wordIndex[word], 0, i);
      //console.log(word, this.wordIndex[word], inputBuffer);
    }
    const input = inputBuffer.toTensor();
    //console.log(input);

    status('Running inference');
    const beginMs = performance.now();
    const predictOut = this.model.predict(input);
    //console.log(predictOut.dataSync());
    const score = predictOut.dataSync();//[0];
    predictOut.dispose();
    const endMs = performance.now();

    return {score: score, elapsed: (endMs - beginMs)};
  }
};

async function setup() {
  if (await urlExists(HOSTED_URLS.model)) {
    status('Model available: ' + HOSTED_URLS.model);
    const button = document.getElementById('load-model');
    button.addEventListener('click', async () => {
      const predictor = await new Classifier().init(HOSTED_URLS);
      prepUI(x => predictor.predict(x));
    });
    button.style.display = 'inline-block';
  }

  status('Standing by.');
}

setup();
"""

In [0]:
with open('index.html','w') as f:
  f.write(index_html)
  
with open('index.js','w') as f:
  f.write(index_js)

In [0]:
!ls

index.html  index.js  model_js


Commit and push everything. Note: we're storing large binary files in GitHub, this isn't ideal (if you want to deploy a model down the road, better to host it in a cloud storage bucket).

In [0]:
!git add . 
!git commit -m "colab -> github"
!git push https://{USER_NAME}:{TOKEN}@github.com/{USER_NAME}/{USER_NAME}.github.io/ master

[master 35bf8c6] colab -> github
 1 file changed, 0 insertions(+), 0 deletions(-)
 rewrite hw4/model_js/group1-shard1of1 (68%)
Counting objects: 5, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 2.82 MiB | 3.00 MiB/s, done.
Total 5 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/Micky-angela/Micky-angela.github.io/
   4d531da..35bf8c6  master -> master


All done! Hopefully everything worked. You may need to wait a few moments for the changes to appear in your site. If not working, check the JavaScript console for errors (in Chrome: View -> Developer -> JavaScript Console).

In [0]:
print("Now, visit https://%s.github.io/%s/" % (USER_NAME, SITE_NAME))

Now, visit https://Micky-angela.github.io/hw4/


If you are debugging and Chrome is failing to pick up your changes, though you've verified they're present in your GitHub repo, see the second answer to: https://superuser.com/questions/89809/how-to-force-refresh-without-cache-in-google-chrome